src/transform/src/analysis/equivalences.rs

```diff
@@ -477,12 +477,148 @@ impl EquivalenceClasses {
+    ///
+    /// This method may produce non-novel equivalences, due to its inability to perform `MSE::reduce`.
+    /// We can end up with e.g. constant expressions that cannot be found until they are so reduced.
+    fn expand(&self) -> Vec<Vec<MirScalarExpr>> {
+        // Consider expanding `self.classes` with novel equivalences.
+        let mut novel = self.implications();
+        for class in novel.iter_mut() {
+            // reduce each expression to its canonical form.
+            for expr in class.iter_mut() {
+                self.remap.reduce_expr(expr);
+            }
+            class.sort();
+            class.dedup();
+            // for a class to be interesting we require at least two elements that do not reference the same root.
+            let common_class = class
+                .iter()
+                .map(|x| self.remap.get(x))
+                .reduce(|prev, this| if prev == this { prev } else { None });
+            if class.len() == 1 || common_class != Some(None) {
+                class.clear();
+            }
+        }
+        novel.retain(|c| !c.is_empty());
+        novel
+    }
+
+    /// Derives potentially novel equivalences without regard for minimization.
+    ///
+    /// This is an opportunity to explore equivalences that do not correspond to expression minimization,
+    /// and therefore should not be used in `minimize_once`. They are still potentially important, but
+    /// required additional guardrails to ensure we reach a fixed point.
+    ///
+    /// The implications will be introduced into `self.classes` and will prompt a round of minimization,
+    /// making it somewhat polite to avoid producing outputs that cannot result in novel equivalences.
+    /// For example, before producing a new equivalence, one could check that the involved terms are not
+    /// already present in the same class.
+    fn implications(&self) -> Vec<Vec<MirScalarExpr>> {
+        let mut new_equivalences = Vec::new();
+
+        // If we see `false == IsNull(foo)` we can add the non-null implications of `foo`.
+        let mut non_null = std::collections::BTreeSet::default();
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_false()) {
+                for e in class.iter() {
+                    if let MirScalarExpr::CallUnary {
+                        func: mz_expr::UnaryFunc::IsNull(_),
+                        expr,
+                    } = e
+                    {
+                        expr.non_null_requirements(&mut non_null);
+                    }
+                }
+            }
+        }
+        // If we see `true == foo` we can add the non-null implications of `foo`.
+        // TODO: generalize to arbitrary non-null, non-error literals; at the moment `true == pred` is
+        // an important idiom to identify for how we express predicates.
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_true()) {
```

Comment:
```suggestion
            if class.iter().any(|c| c.is_literal_ok() && !c.is_literal_null()) {
```

(Or add `MirScalarExpr::is_literal_nonnull()`.)

Reply:
Planning on putting this on a post-merge todo list, to avoid the risk that doing this churns a bunch of plans that we have to investigate.